File: campus-navigate-app/expo-app/src/pdr/pdrPipeline.test.ts
import assert from 'node:assert/strict';
import { test } from 'node:test';

import {
  createTransientMotionBatch,
  DEFAULT_PDR_PIPELINE_CONFIG,
  detectStepCount,
  getRouteMovementDirection,
  rankHeadingCandidates,
  runPdrPipeline,
  shouldMoveForHeading,
} from './model';
import type { MotionInputSample, PdrPipelineState } from './type';

const previousState: PdrPipelineState = {
  headingDegrees: 350,
  timestampMs: 900,
  x: 236,
  y: 648,
};

function sample(
  timestampMs: number,
  headingDegrees: number,
  magnitude: number,
): MotionInputSample {
  return {
    acceleration: { x: magnitude, y: 0, z: 0 },
    headingDegrees,
    timestampMs,
  };
}

test('creates a short transient batch and drops stale or excess samples', () => {
  const samples = [
    sample(700, 0, 1.4),
    sample(910, 2, 1.4),
    sample(940, 5, 1.4),
    sample(970, 8, 1.4),
    sample(995, 10, 1.4),
  ];

  const batch = createTransientMotionBatch({
    config: { ...DEFAULT_PDR_PIPELINE_CONFIG, maxSamplesPerBatch: 3 },
    nowMs: 1000,
    samples,
  });

  assert.deepEqual(
    batch.acceptedSamples.map((acceptedSample) => acceptedSample.timestampMs),
    [940, 970, 995],
  );
  assert.equal(batch.droppedSampleCount, 2);
});

test('ranks desired heading above noisy observed heading when desired is close', () => {
  const candidates = rankHeadingCandidates({
    desiredHeadingDegrees: 0,
    observedHeadingDegrees: 10,
    previousHeadingDegrees: 350,
  });

  assert.equal(candidates[0]?.label, 'desired');
  assert.equal(candidates[0]?.headingDegrees, 0);
});

test('runs PDR summary under the real-time batch latency target', () => {
  const result = runPdrPipeline({
    desiredHeadingDegrees: 0,
    nowMs: 1000,
    previousState,
    samples: [
      sample(920, 7, 0.2),
      sample(940, 8, 0.4),
      sample(960, 10, 1.9),
      sample(990, 12, 1.1),
    ],
  });

  assert.equal(result.acceptedSampleCount, 4);
  assert.equal(result.droppedSampleCount, 0);
  assert.equal(result.latencyMs, 10);
  assert.ok(result.latencyMs <= 100);
  assert.equal(result.estimate.source, 'pdr-summary');
  assert.equal(Math.round(result.estimate.headingDegrees), 9);
  assert.equal(result.estimate.x, 264);
  assert.equal(result.estimate.y, 648);
  assert.equal(result.diagnostics.step.rejectReason, 'ACCEPTED');
  assert.equal(result.diagnostics.step.stepCount, 1);
  assert.equal(result.diagnostics.movement.direction, 'forward');
  assert.equal(result.diagnostics.movement.movedStepCount, 1);
  assert.equal(Math.round(result.nextState.headingDegrees), 9);
  assert.equal(result.nextState.lastStepTimestampMs, 960);
});

test('reports observed facing heading while route-ranked heading controls movement', () => {
  const result = runPdrPipeline({
    desiredHeadingDegrees: 0,
    nowMs: 1000,
    previousState: {
      ...previousState,
      headingDegrees: 60,
    },
    samples: [
      sample(920, 58, 0.2),
      sample(940, 60, 0.4),
      sample(960, 62, 1.9),
      sample(990, 60, 1.1),
    ],
  });

  assert.equal(Math.round(result.estimate.headingDegrees), 60);
  assert.equal(Math.round(result.nextState.headingDegrees), 60);
  assert.equal(result.headingCandidates[0]?.label, 'desired');
  assert.equal(result.estimate.x, 264);
  assert.equal(result.estimate.y, 648);
});

test('moves backward when observed heading faces opposite the route direction', () => {
  const result = runPdrPipeline({
    desiredHeadingDegrees: 0,
    nowMs: 1000,
    previousState: {
      ...previousState,
      backwardConfirmationTimestampMs: 500,
    },
    samples: [
      sample(920, 178, 0.2),
      sample(940, 180, 0.4),
      sample(960, 182, 1.9),
      sample(990, 180, 1.1),
    ],
  });

  assert.equal(Math.round(result.estimate.headingDegrees), 180);
  assert.equal(result.diagnostics.movement.direction, 'backward');
  assert.equal(result.diagnostics.movement.movedStepCount, 1);
  assert.equal(result.estimate.x, 208);
  assert.equal(result.estimate.y, previousState.y);
  assert.equal(result.nextState.lastStepTimestampMs, 960);
});

test('locks movement during startup warm-up while still de-duping the step peak', () => {
  const result = runPdrPipeline({
    desiredHeadingDegrees: 0,
    nowMs: 2000,
    previousState: {
      ...previousState,
      startedAtMs: 500,
    },
    samples: [
      sample(1920, 0, 0.2),
      sample(1940, 0, 0.4),
      sample(1960, 0, 2),
      sample(1990, 0, 1),
    ],
  });

  assert.equal(result.diagnostics.step.rejectReason, 'ACCEPTED');
  assert.equal(result.diagnostics.movement.blockedReason, 'startup-lock');
  assert.equal(result.diagnostics.movement.movedStepCount, 0);
  assert.equal(result.estimate.x, previousState.x);
  assert.equal(result.estimate.y, previousState.y);
  assert.equal(result.nextState.lastStepTimestampMs, 1960);
});

test('blocks weak backward steps to avoid reverse drift while adjusting the phone', () => {
  const result = runPdrPipeline({
    desiredHeadingDegrees: 0,
    nowMs: 4000,
    previousState: {
      ...previousState,
      startedAtMs: 0,
    },
    samples: [
      sample(3920, 180, 0.2),
      sample(3940, 180, 0.4),
      sample(3960, 180, 1.75),
      sample(3990, 180, 0.8),
    ],
  });

  assert.equal(result.diagnostics.step.rejectReason, 'ACCEPTED');
  assert.equal(result.diagnostics.movement.direction, 'backward');
  assert.equal(result.diagnostics.movement.blockedReason, 'backward-weak-step');
  assert.equal(result.diagnostics.movement.movedStepCount, 0);
  assert.equal(result.estimate.x, previousState.x);
  assert.equal(result.nextState.lastStepTimestampMs, 3960);
});

test('requires two strong backward steps before moving backward', () => {
  const first = runPdrPipeline({
    desiredHeadingDegrees: 0,
    nowMs: 4000,
    previousState: {
      ...previousState,
      startedAtMs: 0,
    },
    samples: [
      sample(3920, 180, 0.2),
      sample(3940, 180, 0.4),
      sample(3960, 180, 2.1),
      sample(3990, 180, 0.8),
    ],
  });
  const second = runPdrPipeline({
    desiredHeadingDegrees: 0,
    nowMs: 4600,
    previousState: first.nextState,
    samples: [
      sample(4520, 180, 0.2),
      sample(4540, 180, 0.4),
      sample(4560, 180, 2.1),
      sample(4590, 180, 0.8),
    ],
  });

  assert.equal(first.diagnostics.movement.blockedReason, 'backward-confirming');
  assert.equal(first.diagnostics.movement.movedStepCount, 0);
  assert.equal(second.diagnostics.movement.blockedReason, null);
  assert.equal(second.diagnostics.movement.movedStepCount, 1);
  assert.equal(second.estimate.x, 208);
});

test('blocks movement during shake cooldown after repeated shake spikes', () => {
  const firstShake = runPdrPipeline({
    desiredHeadingDegrees: 0,
    nowMs: 1000,
    previousState: {
      ...previousState,
      startedAtMs: 0,
    },
    samples: [
      sample(920, 0, 0.3),
      sample(960, 0, 6.2),
      sample(990, 0, 0.5),
    ],
  });
  const secondShake = runPdrPipeline({
    desiredHeadingDegrees: 0,
    nowMs: 1400,
    previousState: firstShake.nextState,
    samples: [
      sample(1320, 0, 0.3),
      sample(1360, 0, 6.1),
      sample(1390, 0, 0.5),
    ],
  });
  const blockedStep = runPdrPipeline({
    desiredHeadingDegrees: 0,
    nowMs: 1800,
    previousState: secondShake.nextState,
    samples: [
      sample(1720, 0, 0.2),
      sample(1740, 0, 0.4),
      sample(1760, 0, 2.1),
      sample(1790, 0, 0.9),
    ],
  });
  const afterCooldown = runPdrPipeline({
    desiredHeadingDegrees: 0,
    nowMs: 2800,
    previousState: blockedStep.nextState,
    samples: [
      sample(2720, 0, 0.2),
      sample(2740, 0, 0.4),
      sample(2760, 0, 2.1),
      sample(2790, 0, 0.9),
    ],
  });

  assert.equal(firstShake.diagnostics.step.rejectReason, 'SHAKE_TOO_HIGH');
  assert.equal(secondShake.diagnostics.step.rejectReason, 'SHAKE_TOO_HIGH');
  assert.equal(secondShake.nextState.shakeCooldownUntilMs, 2560);
  assert.equal(blockedStep.diagnostics.step.rejectReason, 'ACCEPTED');
  assert.equal(blockedStep.diagnostics.movement.blockedReason, 'shake-cooldown');
  assert.equal(blockedStep.diagnostics.movement.movedStepCount, 0);
  assert.equal(blockedStep.estimate.x, previousState.x);
  assert.equal(afterCooldown.diagnostics.movement.blockedReason, null);
  assert.equal(afterCooldown.diagnostics.movement.movedStepCount, 1);
});

test('blocks movement briefly when the phone turns in place', () => {
  const result = runPdrPipeline({
    desiredHeadingDegrees: 90,
    nowMs: 4000,
    previousState: {
      ...previousState,
      headingDegrees: 0,
      startedAtMs: 0,
    },
    samples: [
      sample(3920, 90, 0.2),
      sample(3940, 90, 0.4),
      sample(3960, 90, 2.1),
      sample(3990, 90, 0.8),
    ],
  });

  assert.equal(result.diagnostics.step.rejectReason, 'ACCEPTED');
  assert.equal(result.diagnostics.movement.blockedReason, 'turning-in-place');
  assert.equal(result.diagnostics.movement.movedStepCount, 0);
  assert.equal(Math.round(result.estimate.headingDegrees), 90);
  assert.equal(result.estimate.x, previousState.x);
  assert.equal(result.estimate.y, previousState.y);
  assert.equal(result.nextState.turnInPlaceUntilMs, 4700);
});

test('rejects accepted-looking steps caused by phone rotation while standing', () => {
  const result = runPdrPipeline({
    desiredHeadingDegrees: 210,
    nowMs: 4000,
    previousState: {
      ...previousState,
      headingDegrees: 204,
      rotationHeadingSnapshots: [
        { headingDegrees: 190, timestampMs: 3400 },
        { headingDegrees: 204, timestampMs: 3700 },
      ],
      rotationHeadingTravelDegrees: 14,
      startedAtMs: 0,
    },
    samples: [
      sample(3920, 214, 0.5),
      sample(3960, 214, 2.1),
      sample(3990, 214, 1.1),
    ],
  });

  assert.equal(result.diagnostics.step.rejectReason, 'PHONE_ROTATION');
  assert.equal(result.diagnostics.step.stepCount, 0);
  assert.equal(result.diagnostics.step.rotationHeadingTravelDegrees, 24);
  assert.equal(result.diagnostics.movement.movedStepCount, 0);
  assert.equal(result.estimate.x, previousState.x);
  assert.equal(result.nextState.lastStepTimestampMs, undefined);
});

test('keeps real walking turns when average acceleration is walking-like', () => {
  const result = runPdrPipeline({
    desiredHeadingDegrees: 210,
    nowMs: 4000,
    previousState: {
      ...previousState,
      headingDegrees: 204,
      rotationHeadingSnapshots: [
        { headingDegrees: 190, timestampMs: 3400 },
        { headingDegrees: 204, timestampMs: 3700 },
      ],
      rotationHeadingTravelDegrees: 14,
      startedAtMs: 0,
    },
    samples: [
      sample(3920, 214, 0.6),
      sample(3960, 214, 3.4),
      sample(3990, 214, 2),
    ],
  });

  assert.equal(result.diagnostics.step.rejectReason, 'ACCEPTED');
  assert.equal(result.diagnostics.step.rotationHeadingTravelDegrees, 24);
  assert.equal(result.diagnostics.movement.blockedReason, null);
  assert.equal(result.diagnostics.movement.movedStepCount, 1);
});

test('allows forward or backward movement only inside route heading cones', () => {
  assert.equal(
    shouldMoveForHeading({
      config: DEFAULT_PDR_PIPELINE_CONFIG,
      desiredHeadingDegrees: 0,
      observedHeadingDegrees: 45,
    }),
    true,
  );
  assert.equal(
    shouldMoveForHeading({
      config: DEFAULT_PDR_PIPELINE_CONFIG,
      desiredHeadingDegrees: 0,
      observedHeadingDegrees: 180,
    }),
    true,
  );
  assert.equal(
    shouldMoveForHeading({
      config: DEFAULT_PDR_PIPELINE_CONFIG,
      desiredHeadingDegrees: 0,
      observedHeadingDegrees: 90,
    }),
    false,
  );
  assert.equal(
    getRouteMovementDirection({
      config: DEFAULT_PDR_PIPELINE_CONFIG,
      desiredHeadingDegrees: 0,
      observedHeadingDegrees: 180,
    }),
    'backward',
  );
});

test('does not move when no fresh batch samples are accepted', () => {
  const result = runPdrPipeline({
    desiredHeadingDegrees: 0,
    nowMs: 1000,
    previousState,
    samples: [sample(100, 45, 2)],
  });

  assert.equal(result.acceptedSampleCount, 0);
  assert.equal(result.diagnostics.step.rejectReason, 'NO_SAMPLES');
  assert.equal(result.estimate.x, previousState.x);
  assert.equal(result.estimate.y, previousState.y);
});

test('rejects shake-only acceleration without a quiet-to-peak step pattern', () => {
  const result = runPdrPipeline({
    desiredHeadingDegrees: 0,
    nowMs: 1000,
    previousState,
    samples: [
      sample(920, 5, 2.4),
      sample(940, 6, 2.2),
      sample(960, 7, 2.6),
      sample(990, 8, 2.3),
    ],
  });

  assert.equal(result.acceptedSampleCount, 4);
  assert.equal(result.diagnostics.step.rejectReason, 'NO_QUIET_SAMPLE');
  assert.equal(result.estimate.x, previousState.x);
  assert.equal(result.estimate.y, previousState.y);
  assert.equal(result.nextState.lastStepTimestampMs, undefined);
});

test('rejects extreme acceleration spikes as shake instead of walking', () => {
  const result = runPdrPipeline({
    desiredHeadingDegrees: 0,
    nowMs: 1000,
    previousState,
    samples: [
      sample(920, 5, 0.2),
      sample(960, 7, 6),
      sample(990, 8, 0.3),
    ],
  });

  assert.equal(result.diagnostics.step.rejectReason, 'SHAKE_TOO_HIGH');
  assert.equal(result.estimate.x, previousState.x);
  assert.equal(result.estimate.y, previousState.y);
});

test('rejects another step inside the minimum step interval', () => {
  const result = runPdrPipeline({
    desiredHeadingDegrees: 0,
    nowMs: 1000,
    previousState: {
      ...previousState,
      lastStepTimestampMs: 800,
    },
    samples: [
      sample(920, 5, 0.2),
      sample(960, 7, 1.9),
      sample(990, 8, 0.3),
    ],
  });

  assert.equal(result.diagnostics.step.rejectReason, 'TOO_SOON_AFTER_LAST_STEP');
  assert.equal(result.estimate.x, previousState.x);
  assert.equal(result.estimate.y, previousState.y);
});

test('detects at most one step peak per short batch', () => {
  assert.equal(
    detectStepCount({
      config: DEFAULT_PDR_PIPELINE_CONFIG,
      previousState,
      samples: [
        sample(920, 5, 0.2),
        sample(960, 7, 1.9),
        sample(990, 8, 2.1),
      ],
    }),
    1,
  );
});

File: campus-navigate-app/flutter_app/test/domain/pdr/pdr_pipeline_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:indoor_navigation/domain/pdr/algorithms/algorithms.dart';
import 'package:indoor_navigation/domain/pdr/pdr_models.dart';

void main() {
  test('runs PDR summary under the real-time batch latency target', () {
    final result = runPdrPipeline(
      desiredHeadingDegrees: 0,
      nowMs: 1000,
      previousState: _previousState,
      samples: <MotionInputSample>[
        _sample(920, 7, 0.2),
        _sample(940, 8, 0.4),
        _sample(960, 10, 1.9),
        _sample(990, 12, 1.1),
      ],
    );

    expect(result.acceptedSampleCount, 4);
    expect(result.droppedSampleCount, 0);
    expect(result.latencyMs, 10);
    expect(result.latencyMs, lessThanOrEqualTo(100));
    expect(result.estimate.source.wireValue, 'pdr-summary');
    expect(result.estimate.headingDegrees.round(), 9);
    expect(result.estimate.x, 264);
    expect(result.estimate.y, 648);
    expect(result.diagnostics.step.rejectReason, StepRejectReason.accepted);
    expect(result.diagnostics.step.stepCount, 1);
    expect(
      result.diagnostics.movement.direction,
      RouteMovementDirection.forward,
    );
    expect(result.diagnostics.movement.movedStepCount, 1);
    expect(result.nextState.headingDegrees.round(), 9);
    expect(result.nextState.lastStepTimestampMs, 960);
  });

  test(
    'reports facing heading while route-ranked heading controls movement',
    () {
      final result = runPdrPipeline(
        desiredHeadingDegrees: 0,
        nowMs: 1000,
        previousState: _state(headingDegrees: 60),
        samples: <MotionInputSample>[
          _sample(920, 58, 0.2),
          _sample(940, 60, 0.4),
          _sample(960, 62, 1.9),
          _sample(990, 60, 1.1),
        ],
      );

      expect(result.estimate.headingDegrees.round(), 60);
      expect(result.nextState.headingDegrees.round(), 60);
      expect(
        result.headingCandidates.first.label,
        HeadingCandidateLabel.desired,
      );
      expect(result.estimate.x, 264);
      expect(result.estimate.y, 648);
    },
  );

  test('moves backward when facing opposite the route direction', () {
    final result = runPdrPipeline(
      desiredHeadingDegrees: 0,
      nowMs: 1000,
      previousState: _state(backwardConfirmationTimestampMs: 500),
      samples: _stepSamples(nowMs: 1000, headingDegrees: 180, peak: 1.9),
    );

    expect(result.estimate.headingDegrees.round(), 180);
    expect(
      result.diagnostics.movement.direction,
      RouteMovementDirection.backward,
    );
    expect(result.diagnostics.movement.movedStepCount, 1);
    expect(result.estimate.x, 208);
    expect(result.estimate.y, _previousState.y);
    expect(result.nextState.lastStepTimestampMs, 960);
  });

  test('locks movement during startup while de-duplicating the peak', () {
    final result = runPdrPipeline(
      desiredHeadingDegrees: 0,
      nowMs: 2000,
      previousState: _state(startedAtMs: 500),
      samples: _stepSamples(nowMs: 2000, headingDegrees: 0, peak: 2),
    );

    expect(result.diagnostics.step.rejectReason, StepRejectReason.accepted);
    expect(
      result.diagnostics.movement.blockedReason,
      MovementBlockedReason.startupLock,
    );
    expect(result.diagnostics.movement.movedStepCount, 0);
    expect(result.estimate.x, _previousState.x);
    expect(result.nextState.lastStepTimestampMs, 1960);
  });

  test('blocks weak backward steps', () {
    final result = runPdrPipeline(
      desiredHeadingDegrees: 0,
      nowMs: 4000,
      previousState: _state(startedAtMs: 0),
      samples: _stepSamples(nowMs: 4000, headingDegrees: 180, peak: 1.75),
    );

    expect(result.diagnostics.step.rejectReason, StepRejectReason.accepted);
    expect(
      result.diagnostics.movement.blockedReason,
      MovementBlockedReason.backwardWeakStep,
    );
    expect(result.estimate.x, _previousState.x);
    expect(result.nextState.lastStepTimestampMs, 3960);
  });

  test('requires two strong backward steps before movement', () {
    final first = runPdrPipeline(
      desiredHeadingDegrees: 0,
      nowMs: 4000,
      previousState: _state(startedAtMs: 0),
      samples: _stepSamples(nowMs: 4000, headingDegrees: 180, peak: 2.1),
    );
    final second = runPdrPipeline(
      desiredHeadingDegrees: 0,
      nowMs: 4600,
      previousState: first.nextState,
      samples: _stepSamples(nowMs: 4600, headingDegrees: 180, peak: 2.1),
    );

    expect(
      first.diagnostics.movement.blockedReason,
      MovementBlockedReason.backwardConfirming,
    );
    expect(first.diagnostics.movement.movedStepCount, 0);
    expect(second.diagnostics.movement.blockedReason, isNull);
    expect(second.diagnostics.movement.movedStepCount, 1);
    expect(second.estimate.x, 208);
  });

  test('blocks movement during cooldown after repeated shake spikes', () {
    final firstShake = runPdrPipeline(
      desiredHeadingDegrees: 0,
      nowMs: 1000,
      previousState: _state(startedAtMs: 0),
      samples: _shakeSamples(nowMs: 1000, peak: 6.2),
    );
    final secondShake = runPdrPipeline(
      desiredHeadingDegrees: 0,
      nowMs: 1400,
      previousState: firstShake.nextState,
      samples: _shakeSamples(nowMs: 1400, peak: 6.1),
    );
    final blockedStep = runPdrPipeline(
      desiredHeadingDegrees: 0,
      nowMs: 1800,
      previousState: secondShake.nextState,
      samples: _stepSamples(nowMs: 1800, headingDegrees: 0, peak: 2.1),
    );
    final afterCooldown = runPdrPipeline(
      desiredHeadingDegrees: 0,
      nowMs: 2800,
      previousState: blockedStep.nextState,
      samples: _stepSamples(nowMs: 2800, headingDegrees: 0, peak: 2.1),
    );

    expect(
      firstShake.diagnostics.step.rejectReason,
      StepRejectReason.shakeTooHigh,
    );
    expect(
      secondShake.diagnostics.step.rejectReason,
      StepRejectReason.shakeTooHigh,
    );
    expect(secondShake.nextState.shakeCooldownUntilMs, 2560);
    expect(
      blockedStep.diagnostics.movement.blockedReason,
      MovementBlockedReason.shakeCooldown,
    );
    expect(blockedStep.diagnostics.movement.movedStepCount, 0);
    expect(afterCooldown.diagnostics.movement.blockedReason, isNull);
    expect(afterCooldown.diagnostics.movement.movedStepCount, 1);
  });

  test('blocks movement briefly when the phone turns in place', () {
    final result = runPdrPipeline(
      desiredHeadingDegrees: 90,
      nowMs: 4000,
      previousState: _state(headingDegrees: 0, startedAtMs: 0),
      samples: _stepSamples(nowMs: 4000, headingDegrees: 90, peak: 2.1),
    );

    expect(result.diagnostics.step.rejectReason, StepRejectReason.accepted);
    expect(
      result.diagnostics.movement.blockedReason,
      MovementBlockedReason.turningInPlace,
    );
    expect(result.diagnostics.movement.movedStepCount, 0);
    expect(result.estimate.headingDegrees.round(), 90);
    expect(result.nextState.turnInPlaceUntilMs, 4700);
  });

  test('rejects accepted-looking steps caused by phone rotation', () {
    final result = runPdrPipeline(
      desiredHeadingDegrees: 210,
      nowMs: 4000,
      previousState: _state(
        headingDegrees: 204,
        rotationHeadingSnapshots: <PdrHeadingSnapshot>[
          const PdrHeadingSnapshot(headingDegrees: 190, timestampMs: 3400),
          const PdrHeadingSnapshot(headingDegrees: 204, timestampMs: 3700),
        ],
        rotationHeadingTravelDegrees: 14,
        startedAtMs: 0,
      ),
      samples: <MotionInputSample>[
        _sample(3920, 214, 0.5),
        _sample(3960, 214, 2.1),
        _sample(3990, 214, 1.1),
      ],
    );

    expect(
      result.diagnostics.step.rejectReason,
      StepRejectReason.phoneRotation,
    );
    expect(result.diagnostics.step.stepCount, 0);
    expect(result.diagnostics.step.rotationHeadingTravelDegrees, 24);
    expect(result.diagnostics.movement.movedStepCount, 0);
    expect(result.nextState.lastStepTimestampMs, isNull);
  });

  test('keeps walking turns when average acceleration is walking-like', () {
    final result = runPdrPipeline(
      desiredHeadingDegrees: 210,
      nowMs: 4000,
      previousState: _state(
        headingDegrees: 204,
        rotationHeadingSnapshots: <PdrHeadingSnapshot>[
          const PdrHeadingSnapshot(headingDegrees: 190, timestampMs: 3400),
          const PdrHeadingSnapshot(headingDegrees: 204, timestampMs: 3700),
        ],
        startedAtMs: 0,
      ),
      samples: <MotionInputSample>[
        _sample(3920, 214, 0.6),
        _sample(3960, 214, 3.4),
        _sample(3990, 214, 2),
      ],
    );

    expect(result.diagnostics.step.rejectReason, StepRejectReason.accepted);
    expect(result.diagnostics.step.rotationHeadingTravelDegrees, 24);
    expect(result.diagnostics.movement.blockedReason, isNull);
    expect(result.diagnostics.movement.movedStepCount, 1);
  });

  test('does not move with empty, stale, or future-only input', () {
    for (final samples in <List<MotionInputSample>>[
      <MotionInputSample>[],
      <MotionInputSample>[_sample(100, 45, 2)],
      <MotionInputSample>[_sample(1010, 45, 2)],
    ]) {
      final result = runPdrPipeline(
        desiredHeadingDegrees: 0,
        nowMs: 1000,
        previousState: _previousState,
        samples: samples,
      );

      expect(result.acceptedSampleCount, 0);
      expect(result.diagnostics.step.rejectReason, StepRejectReason.noSamples);
      expect(result.estimate.x, _previousState.x);
      expect(result.estimate.y, _previousState.y);
    }
  });

  test('sorts out-of-order input before step analysis', () {
    final result = runPdrPipeline(
      desiredHeadingDegrees: 0,
      nowMs: 1000,
      pixelsPerMeter: 20,
      previousState: _state(headingDegrees: 0),
      samples: <MotionInputSample>[_sample(950, 0, 1.9), _sample(900, 0, 0.2)],
    );

    expect(result.diagnostics.step.rejectReason, StepRejectReason.accepted);
    expect(result.diagnostics.batch.sampleStartTimestampMs, 900);
    expect(result.diagnostics.batch.sampleEndTimestampMs, 950);
    expect(result.diagnostics.movement.distancePixels, 10);
  });
}

final _previousState = PdrPipelineState(
  headingDegrees: 350,
  timestampMs: 900,
  x: 236,
  y: 648,
);

MotionInputSample _sample(
  int timestampMs,
  double headingDegrees,
  double magnitude,
) {
  return MotionInputSample(
    acceleration: MotionVector(x: magnitude, y: 0, z: 0),
    headingDegrees: headingDegrees,
    timestampMs: timestampMs,
  );
}

List<MotionInputSample> _stepSamples({
  required int nowMs,
  required double headingDegrees,
  required double peak,
}) {
  return <MotionInputSample>[
    _sample(nowMs - 80, headingDegrees, 0.2),
    _sample(nowMs - 60, headingDegrees, 0.4),
    _sample(nowMs - 40, headingDegrees, peak),
    _sample(nowMs - 10, headingDegrees, 0.8),
  ];
}

List<MotionInputSample> _shakeSamples({
  required int nowMs,
  required double peak,
}) {
  return <MotionInputSample>[
    _sample(nowMs - 80, 0, 0.3),
    _sample(nowMs - 40, 0, peak),
    _sample(nowMs - 10, 0, 0.5),
  ];
}

PdrPipelineState _state({
  int? backwardConfirmationTimestampMs,
  double? headingDegrees,
  List<PdrHeadingSnapshot>? rotationHeadingSnapshots,
  double? rotationHeadingTravelDegrees,
  int? startedAtMs,
}) {
  return PdrPipelineState(
    backwardConfirmationTimestampMs: backwardConfirmationTimestampMs,
    headingDegrees: headingDegrees ?? _previousState.headingDegrees,
    rotationHeadingSnapshots: rotationHeadingSnapshots,
    rotationHeadingTravelDegrees: rotationHeadingTravelDegrees,
    startedAtMs: startedAtMs,
    timestampMs: _previousState.timestampMs,
    x: _previousState.x,
    y: _previousState.y,
  );
}
